Add monetary balance verification

rtsisyk opened the pull request:
- Add bulletproofs validation
- Validate tx monetary balance explicitly (was implicit via signature)
- Validate block monetary balance

Closes #178
Closes #191

rtsisyk commented on blockchain/src/block.rs:
```diff
@@ -243,6 +249,49 @@ impl MonetaryBlock {
+
+            let pedersen_commitment: Pt = txin.proof.pedersen_commitment();
+            let pedersen_commitment: ECp = Pt::decompress(pedersen_commitment)?;
+            gamma += pedersen_commitment;
```
I actualy use ECp here:

```
      let mut gamma = ECp::inf();

     for .. {
            let pedersen_commitment: Pt = txin.proof.pedersen_commitment();
            let pedersen_commitment: ECp = Pt::decompress(pedersen_commitment)?;
            gamma += pedersen_commitment;
        }

        if gamma != self.header.adjustment * (*G) {
            return Err(BlockchainError::InvalidBlockBalance.into());
        }
```

self.header.adjustment is Fr of course.

dbmcclain replied:
okay... your notion of gamma is not the traditional interpretation. You are just looking at the sum of Pedersen Commitments.

Unfortunately, anyone can do that, and it proves nothing. Gamma is the scalar sum of all the gamma factors from TXIN's, minus the gamma factors from TXOUT's, minus the gamma factor used for the fee.

Only someone with complete knowledge of the monetary amounts and gamma factors used would be able to tell you what value you should use for gamma_adj.

If they can do so, it proves that they have the necessary ownership knowledge. If they can't do so, it means they must be trying to fake it by simply adding up the Pedersen commitments

rtsisyk replied:
I'm checking here that the sum of all pedersent commitments for inputs minus the sum of pedersen commitments for outputs minus fee is equal to block/tx gamma adjustment. 

But, stop.. Do you mean that someone can just create transaction/block with special fake value of pedersen commitment and/or gamma_adjustment in order to bypass this  check?

rtsisyk replied:
OK, I understood what did you mean. Anyone can form a transaction with special fake value of pedersen commitment in order to satisfy sum(C_i) - sum(C_j) == gamma_adjustment * G. In this case my patch is almost useless...

It seems that we have to send original transactions with boundaries in initial CoSi round...

rtsisyk replied:
I renamed gamma to `pedersen_commitment_diff`. I think now it is more clear.

rtsisyk commented:
> You are performing the correct verification computation, but I must take issue with your renaming departures from the yellowpaper. The sum/ difference you are doing is not gamma, but rather a commitment sum / difference. You check that the result equals block_gamma_adj * G. That is correct.

I agree that name `gamma` here is not appropriate. I'll rename to commitment sum / difference.

> if you were to accept UTXO's using this blind acceptance of gamma_adj, as computed by the sum of Pedersen commitments, then you risk accepting totally fictitious amounts into the blockchain. People could manufacture riches for themselves.

These UTXO should be checked during transaction validation.

Please don't merge this branch right now, I want to rework it a little bit.

rtsisyk commented:
- Renamed `gamma` in both monetary balance checks to `pedersen_commitment_diff` to make it more clear
- Renamed block.adjustment to block.gamma to match tx.gamma
- Fix all places where name `delta` was used by mistake when called Output::new()
- Re-based on the latest dev after rust2018 patch
